File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
input = sys.stdin.readline

n = int(input())
grid = [list(map(int, input().split())) for _ in range(n)]

num = [i for i in range(n)]
def comb(arr, r):
  for i in range(len(arr)):
    if r == 1:
      yield [arr[i]]
    else:
      for next in comb(arr[i+1:], r-1):
        yield [arr[i]]+next

starts = []
# 팀 추출
for arr in comb(num, n/2):
  starts.append(arr)
# 능력치 계산
ans = int(1e9)
# 능력치는 스타트팀(Sij + Sji) - 링크팀(Si'j' + Sj'i')
# 자기 자신은 0 이니깐 계산에 영향 x
for start in starts:
  teamStart = 0
  teamLink = 0
  for i in start:
    for j in start:
      teamStart += grid[i][j]
  link = list(set(num)-set(start))
  for i in link:
    for j in link:
      teamLink += grid[i][j]
  ans = min(ans, abs(teamStart-teamLink))

print(ans)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static int n, total = 0;
	static int[][] s;
	static int[] numbers;
	static ArrayList<ArrayList<Integer>> starts;
	static boolean[] visited;
	
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;
        
        n = Integer.parseInt(br.readLine());
        s = new int[n+1][n+1];
        
        for (int i =1; i<n+1; i++) {
        	st = new StringTokenizer(br.readLine());
        	for (int j=1; j<n+1; j++) {
        		s[i][j] = Integer.parseInt(st.nextToken());
        		
        		total += s[i][j];
        	}
        }
        
        numbers = new int[n];
        for (int i =0; i<n; i++) {
        	numbers[i] = i+1;
        }
        
        visited = new boolean[n+1];
        starts = new ArrayList<>();
        combination(0, 0, n / 2);
        
        int ans = total;
        for (ArrayList<Integer> start : starts) {
        	int teamStart = 0;
            int teamLink = 0;
        	
        	for (int num1 : start) {
        		for (int num2 : start) {
        			teamStart += s[num1][num2];
        		}
        	}
        	
            ArrayList<Integer> link = new ArrayList<>();
            for (int i = 1; i <= n; i++) {
                if (!start.contains(i)) {
                    link.add(i);
                }
            }

            for (int num1 : link) {
                for (int num2 : link) {
                    teamLink += s[num1][num2];
                }
            }
        	int diff = Math.abs(teamStart - teamLink);
        	
        	if (diff < ans) {
        		ans = Math.min(diff, ans);
        	}
        }
        
        System.out.print(ans);
    }
    
    private static void combination(int depth, int start, int r) {
        if (r == 0) {
            ArrayList<Integer> team = new ArrayList<>();
            for (int i=0; i<n+1; i++) {
            	if (visited[i]) team.add(numbers[i]);
            }
            
            starts.add(team);
            return;
        }

        for (int i = start; i < n; i++) {
            visited[i] = true;
            combination(depth + 1, i + 1, r - 1);
            visited[i] = false;
        }
    }

}